File: lib/app/domain/movement_queue.dart
import 'movement_log.dart';
import 'movement_log_sink.dart';
import 'movement_transport.dart';
import 'sync_options.dart';
import 'sync_queue.dart';
import 'sync_trigger.dart';

class MovementQueue implements SyncQueue, MovementLogSink {
  MovementQueue({required MovementTransport transport})
      : _transport = transport;

  final MovementTransport _transport;
  final Map<String, MovementLog> _logs = {};

  int totalCount() => _logs.length;

  MovementLog? getById(String id) => _logs[id];

  List<MovementLog> listUnsynced() {
    return _logs.values.where((item) => !item.isSynced).toList(growable: false);
  }

  int unsyncedCount() => listUnsynced().length;

  Future<MovementLog> enqueue(MovementLog movement) async {
    final existing = _logs[movement.id];
    if (existing != null) {
      return existing;
    }
    _logs[movement.id] = movement;
    return movement;
  }

  @override
  Future<MovementLog> saveLog(MovementLog movement) async {
    final queued = await enqueue(movement.copyWith(isSynced: false));
    if (queued.isSynced) {
      return queued;
    }

    try {
      await _transport.send(
        [queued],
        const SyncOptions(
          isOfflineSync: false,
          trigger: SyncTrigger.immediate,
        ),
      );
      final synced = queued.copyWith(isSynced: true);
      _logs[queued.id] = synced;
      return synced;
    } catch (_) {
      final unsynced = queued.copyWith(isSynced: false);
      _logs[queued.id] = unsynced;
      return unsynced;
    }
  }

  @override
  Future<int> syncPending({SyncTrigger trigger = SyncTrigger.manual}) async {
    final pending = listUnsynced();
    if (pending.isEmpty) {
      return 0;
    }

    try {
      await _transport.send(
        pending,
        SyncOptions(
          isOfflineSync: true,
          trigger: trigger,
        ),
      );

      for (final item in pending) {
        _logs[item.id] = item.copyWith(isSynced: true);
      }
      return pending.length;
    } catch (_) {
      return 0;
    }
  }
}

File: src/app/movement_queue.js
class MovementQueue {
  constructor({ transport }) {
    if (!transport || typeof transport.send !== 'function') {
      throw new Error('transport.send is required');
    }

    this.transport = transport;
    this.logs = new Map();
  }

  totalCount() {
    return this.logs.size;
  }

  getById(id) {
    return this.logs.get(id) ?? null;
  }

  listUnsynced() {
    return Array.from(this.logs.values()).filter((item) => item.isSynced === false);
  }

  unsyncedCount() {
    return this.listUnsynced().length;
  }

  async enqueue(movement) {
    const existing = this.logs.get(movement.id);
    if (existing) {
      return existing;
    }

    const normalized = {
      ...movement,
      isSynced: movement.isSynced ?? false
    };
    this.logs.set(normalized.id, normalized);
    return normalized;
  }

  _markSynced(batch) {
    for (const item of batch) {
      const existing = this.logs.get(item.id);
      if (existing) {
        existing.isSynced = true;
      }
    }
  }

  async saveLog(movement) {
    const log = await this.enqueue(movement);

    if (log.isSynced) {
      return log;
    }

    try {
      await this.transport.send([log], {
        isOfflineSync: false,
        trigger: 'immediate'
      });
      log.isSynced = true;
    } catch (_error) {
      log.isSynced = false;
    }

    return log;
  }

  async syncPending({ trigger = 'manual' } = {}) {
    const pending = this.listUnsynced();
    if (pending.length === 0) {
      return 0;
    }

    try {
      await this.transport.send(pending, {
        isOfflineSync: true,
        trigger
      });
      this._markSynced(pending);
      return pending.length;
    } catch (_error) {
      return 0;
    }
  }
}

module.exports = {
  MovementQueue
};
